### PR TITLE
DTMF-additional-support - add support for IN_BAND DTMF

### PR DESCRIPTION
**Pull request by acharlop**
based off of comments for #496 and #601

### CODE REVIEW

**Comment by acharlop** [comment]
See documentation PR as well https://github.com/versatica/jssip.net/pull/5

**Comment by jmillan** [comment]
It is based on this PR, yes? https://github.com/versatica/JsSIP/pull/601

File: lib/RTCSession.js [hunk 1]
```diff
@@ -942,6 +942,7 @@ module.exports = class RTCSession extends EventEmitter
     let position = 0;
     let duration = options.duration || null;
     let interToneGap = options.interToneGap || null;
+    const transportType = options.transportType || JsSIP_C.DTMF_TRANSPORT.INFO;
```
Comment (ibc) [inline lib/RTCSession.js:945]: After line 945, please validate `transportType`. If not a valid one, throw a `TypeError`.

File: lib/RTCSession.js [hunk 2]
```diff
@@ -1046,13 +1047,32 @@ module.exports = class RTCSession extends EventEmitter
       }
       else
       {
-        const dtmf = new RTCSession_DTMF(this);
+        // send DTMF according to transport config
+        switch (transportType)
```
Comment (ibc) [inline lib/RTCSession.js:1051]: Comments always with proper capital letter and dot at the end.

```suggestion

        // Send DTMF according to transport config.
        switch (transportType)
```

File: lib/RTCSession.js [hunk 3]
```diff
@@ -1046,13 +1047,32 @@ module.exports = class RTCSession extends EventEmitter
...
+        {
+          case JsSIP_C.DTMF_TRANSPORT.IN_BAND: {
+            // send DTMF in current audio RTP stream
```
Comment (ibc) [inline lib/RTCSession.js:1054]:
```suggestion
            // Send DTMF in current audio RTP stream.
```

File: lib/RTCSession.js [hunk 4]
```diff
@@ -1046,13 +1047,32 @@ module.exports = class RTCSession extends EventEmitter
...
 
-        options.eventHandlers = {
-          onFailed : () => { this._tones = null; }
-        };
-        dtmf.send(tone, options);
-        timeout = duration + interToneGap;
+            break;
+          }
+          case JsSIP_C.DTMF_TRANSPORT.INFO:
+          default: {
```
Comment (ibc) [inline lib/RTCSession.js:1065]: Don't use `default`. `transportType` must be validated above.

```suggestion
          case JsSIP_C.DTMF_TRANSPORT.INFO:
          {
```

**Comment by acharlop** [comment]
> It is based on this PR, yes? #601

See PR description both #601 and the original #496

File: lib/RTCSession.js [hunk 5]
```diff
@@ -1046,13 +1047,32 @@ module.exports = class RTCSession extends EventEmitter
...
+            // send DTMF via SIP INFO messages
+            const dtmf = new RTCSession_DTMF(this);
+
+            options.eventHandlers = {
```
Comment (ibc) [inline lib/RTCSession.js:1069]: Does this really pass `gulp lint` task?

File: lib/RTCSession.js [hunk 6]
```diff
@@ -2632,6 +2652,33 @@ module.exports = class RTCSession extends EventEmitter
       });
   }
 
+  /**
+   * DTMF RTP Sender
```
Comment (ibc) [inline lib/RTCSession.js:2656]:
```suggestion
   * Get DTMF RTCRtpSender.
```

File: lib/RTCSession.js (hunk 5 above)
Comment (acharlop) [reply lib/RTCSession.js:1069]: yes

File: lib/RTCSession.js [hunk 7]
```diff
@@ -2632,6 +2652,33 @@ module.exports = class RTCSession extends EventEmitter
...
+  _getDTMFRTPSender()
+  {
+    const senders = this._connection.getSenders().find((sender) =>
+    {
+      return sender.track && sender.track.kind === 'audio';
+    });
+
+    if (!senders)
+    {
+      debugerror('No local audio track to send DTMF with');
```
Comment (ibc) [inline lib/RTCSession.js:2667]: We don't use this kind of logs. Please, keep the JsSIP code style. Same below.

Comment (acharlop) [reply lib/RTCSession.js:2667]: can you please clarify. Do you mean it should be `sendDTMF() | No local audio track to send DTMF with`

Comment (ibc) [reply lib/RTCSession.js:2667]: `sendDTMF() | no local audio track to send DTMF with`

File: lib/RTCSession.js (hunk 1 above)
Comment (acharlop) [reply lib/RTCSession.js:945]: done

**Comment by acharlop** [comment]
Should be g2g now

File: lib/Constants.js [hunk 8]
```diff
@@ -64,6 +64,12 @@ module.exports = {
   UPDATE    : 'UPDATE',
   SUBSCRIBE : 'SUBSCRIBE',
 
+  // DTMF transport methods
```
Comment (ibc) [inline lib/Constants.js:67]:
```suggestion
  // DTMF transport methods.
```

File: lib/RTCSession.js [hunk 9]
```diff
@@ -1046,13 +1047,36 @@ module.exports = class RTCSession extends EventEmitter
       }
       else
       {
-        const dtmf = new RTCSession_DTMF(this);
+        if (transportType !== JsSIP_C.DTMF_TRANSPORT.INFO ||
+          transportType !== JsSIP_C.DTMF_TRANSPORT.IN_BAND)
+        {
+          throw new TypeError(`invalid transportType: ${transportType}`);
+        }
+        // Send DTMF according to transport config.
```
Comment (ibc) [inline lib/RTCSession.js:1055]:
```suggestion
        }

        // Send DTMF according to transport config.
```

File: lib/RTCSession.js [hunk 10]
```diff
@@ -1046,13 +1047,36 @@ module.exports = class RTCSession extends EventEmitter
...
+            }
 
-        options.eventHandlers = {
-          onFailed : () => { this._tones = null; }
-        };
-        dtmf.send(tone, options);
-        timeout = duration + interToneGap;
+            break;
+          }
+          case JsSIP_C.DTMF_TRANSPORT.INFO: {
```
Comment (ibc) [inline lib/RTCSession.js:1069]: Blank line between `case`. BTW: does hit pass `gulp lint` or not? You did not answer before.

```suggestion
          }

          case JsSIP_C.DTMF_TRANSPORT.INFO: {
```

File: lib/RTCSession.js [hunk 11]
```diff
@@ -2632,6 +2656,26 @@ module.exports = class RTCSession extends EventEmitter
       });
   }
 
+  /**
+   * Get DTMF RTCRtpSender.
+   */
+  _getDTMFRTPSender()
+  {
+    const senders = this._connection.getSenders().find((sender) =>
```
Comment (ibc) [inline lib/RTCSession.js:2664]: This is not `senders` but `sender`, right?

```suggestion
    const sender = this._connection.getSenders().find((sender) =>
```

File: lib/RTCSession.js (hunk 10 above)
Comment (acharlop) [reply lib/RTCSession.js:1069]: yes it does pass. I will add the blank line as requested

File: lib/RTCSession.js (hunk 11 above)
Comment (acharlop) [reply lib/RTCSession.js]: done

File: lib/Constants.js (hunk 8 above)
Comment (acharlop) [reply lib/Constants.js]: done

File: lib/RTCSession.js (hunk 9 above)
Comment (acharlop) [reply lib/RTCSession.js]: done

**Comment by acharlop** [comment]
> What about this line which is **also** called even when using in band DTMF?:

Isn't this just for handling if multiple tones are sent? What would the problem be?

**Comment by rshpount** [comment]
I think the terminology here is wrong (or confusing). For instance https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Using_DTMF refers to in-band as sending DTMF as an audio signal using G.711. RFC4733 tones, which are generated by RTCDTMFSender.insertDTMF, are referred to as out of band. SIP INFO messages are technically out-of-band as well. Better name, instead of IN_BAND, should be something like TELEPHONE_EVENT or RFC2833 or RFC4733.

**Comment by ibc** [comment]
Right, thanks @rshpount. 
@acharlop please, rename the type to `TELEPHONE_EVENT`.

Regarding "multiple tones" you are right, just let's wait for @jmillan to jump into this again and confirm.

**Comment by acharlop** [comment]
Since this PR is based off of #601 I believe `RFC2833` would be the correct terminology to use.

File: lib/RTCSession.js [hunk 12]
```diff
@@ -1046,13 +1047,38 @@ module.exports = class RTCSession extends EventEmitter
       }
       else
       {
-        const dtmf = new RTCSession_DTMF(this);
+        if (transportType !== JsSIP_C.DTMF_TRANSPORT.INFO ||
+          transportType !== JsSIP_C.DTMF_TRANSPORT.RFC2833)
+        {
```
Comment (ibc) [inline lib/RTCSession.js:1052]: cosmetic:

```suggestion
        if (
          transportType !== JsSIP_C.DTMF_TRANSPORT.INFO ||
          transportType !== JsSIP_C.DTMF_TRANSPORT.RFC2833
        )
        {
```

Comment (ibc) [reply lib/RTCSession.js:1052]: please commit suggestion directly by clicking into "Commit suggestion" button.

**Comment by ibc** [comment]
Just a last minor request change and we are ready to merge.

**Comment by jmillan** [comment]
```js
// Set timeout for the next tone.
setTimeout(_sendDTMF.bind(this), timeout);
```

> @jmillan what to do here?


It's OK, since we are sending a DTMF at a time in both cases.

**Comment by acharlop** [comment]
> Just a last minor request change and we are ready to merge.

@ibc done

**Comment by ibc** [comment]
Thanks, shipped in JsSIP 3.4.5.

**Comment by ibc** [comment]
Hi, @acharlop. Did you really test your code? It was not working due to a clear bug:

https://github.com/versatica/JsSIP/issues/630

Fixing it now.

**Pull request merged**